program/src/token.rs

```diff
@@ -69,11 +69,18 @@ macro_rules! impl_token {
 
         impl fmt::Display for $TokenLamports {
             fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-                write!(
-                    f,
-                    "{}.{:0>9} {}",
+                let str_number = format!(
+                    "{}.{:0>9}",
                     self.0 / 10u64.pow($decimals),
                     self.0 % 10u64.pow($decimals),
+                );
+                write!(
+                    f,
+                    "{} {}",
+                    &str_number[..str_number.len() - (9 - $decimals)],
```

Comment:
This does not look right to me. If we have fewer decimals, we should cut out the zeros after the decimal dot, not the zeros at the end.

Please add a unit test for this to verify that the output is correct. E.g.

```rust
#[test]
fn test_token_format() {
    assert_eq!(format!("{}", Lamports(1)), "0.000000001 SOL");
    assert_eq!(format!("{}", Lamports(1_000_000_002)), "1.000000002 SOL");
    assert_eq!(format!("{}", UstLamports(1)), "0.000001 UST");
    assert_eq!(format!("{}", UstLamports(1_000_000_002)), "1000.000002 UST");
}
```

Also, now that I am writing this, I realize that we need to replace the `let mut exponent = 9_i32` in the `FromStr` impl with `let mut exponent: i32 = $decimals;`. There is a test `test_lamports_from_str_roundtrip` already. We can extend this test to test `UstLamports` in addition to `Lamports`. (Also, maybe the term “lamports” is confusing here when it doesn’t mean 10<sup>-9</sup>, we could call it `MicroUst` or something.)